relay-sampling/src/evaluation.rs

```diff
@@ -95,139 +114,193 @@ pub fn merge_configs_and_match(
     // The determination of the sampling mode occurs with the following priority:
     // 1. Non-root project sampling mode
     // 2. Root project sampling mode
-    let Some(primary_config) = sampling_config.or(root_sampling_config) else {
-        relay_log::error!("cannot sample without at least one sampling config");
-        return None;
-    };
-    let sample_rate = match primary_config.mode {
-        SamplingMode::Received => match_result.sample_rate,
-        SamplingMode::Total => match dsc {
-            Some(dsc) => dsc.adjusted_sample_rate(match_result.sample_rate),
-            None => match_result.sample_rate,
-        },
-        SamplingMode::Unsupported => {
+    let adjust_sample_rate = match sampling_config
+        .or(root_sampling_config)
+        .map(|config| config.mode)
+    {
+        Some(SamplingMode::Total) => true,
+        Some(SamplingMode::Received) => false,
+        Some(SamplingMode::Unsupported) => {
             if processing_enabled {
                 relay_log::error!("found unsupported sampling mode even as processing Relay");
             }
-
-            return None;
+            return SamplingResult::NoMatch;
+        }
+        None => {
+            relay_log::info!("cannot sample without at least one sampling config");
+            return SamplingResult::NoMatch;
         }
     };
-    match_result.set_sample_rate(sample_rate);
 
-    // Only if we arrive at this stage, it means that we have found a match and we want to prepare
-    // the data for making the sampling decision.
-    Some(match_result)
+    // We perform the rule matching with the multi-matching logic on the merged rules.
+    let Some(rules) =
+        verify_and_merge_rules(processing_enabled, sampling_config, root_sampling_config)
+    else {
+        return SamplingResult::NoMatch;
+    };
+
+    get_sampling_match(rules, event, dsc, now, adjust_sample_rate)
 }
 
 /// Represents the specification for sampling an incoming event.
-#[derive(Clone, Debug, PartialEq, Serialize)]
-pub struct SamplingMatch {
-    /// The sample rate to use for the incoming event.
-    pub sample_rate: f64,
+#[derive(Default, Clone, Debug, PartialEq)]
+pub enum SamplingResult {
+    /// The event matched a sampling condition.
+    Match {
```

Comment:
@jan-auer  had 2nd thoughts on creating a named struct here, where it's used the anonymous one seems just as good

Reply:
It's less convenient this way. You can see the effects in usage sites even in this PR already. Named structs can have their own methods, can be passed to functions, can have private fields (which this one should) and can be matched with a name. 

For instance, instead of writing the following:

```rust
match result {
  SamplingResult::Match {
    is_kept,
    matched_rules,
    ..
  } if is_kept => { ... }
}
```

... one can write this:

```rust
match result {
  SamplingResult::Match(m) if m.should_keep() => handle_match(m),
  ...
}
```

Note how in the first example it would be impossible to write a function that takes the match. Either you'd have to pass all fields, or you'd have to match again in the function and unwrap.

Additionally, I'd ask to make all fields private and add convenience methods.

Reply:
done :)